Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,552 @@
+    return evalComp (← NormalExpr.associator (.of f) g (.of h)) (evalComp η' (evalComp (← NormalExpr.associatorInv (.of f) g' (.of h)) θ'))
+  | η, .comp g h => do
+    let η' ← evalWhiskerRightExpr (← evalWhiskerRightExpr η g) h
+    let f ← η.src
+    let f' ← η.tar
+    return evalComp (← NormalExpr.associatorInv f g h) (evalComp η' (← NormalExpr.associator f' g h))
+  | η, .id _ => do
+    let f ← η.src
+    let g ← η.tar
+    return evalComp (← NormalExpr.rightUnitor f) (evalComp η (← NormalExpr.rightUnitorInv g))
+
+def WhiskerRightExpr.e : WhiskerRightExpr → MetaM Expr
+  | WhiskerRightExpr.of η => η.e
+  | WhiskerRightExpr.whisker η f => do
+    mkAppM ``MonoidalCategoryStruct.whiskerRight #[← η.e, f.e]
+
+def WhiskerLeftExpr.e : WhiskerLeftExpr → MetaM Expr
+  | WhiskerLeftExpr.of η => η.e
+  | WhiskerLeftExpr.whisker f η => do
+    mkAppM ``MonoidalCategoryStruct.whiskerLeft #[f.e, ← η.e]
+
+def NormalExpr.e : NormalExpr → MetaM Expr
+  | NormalExpr.id f => do mkAppM ``CategoryStruct.id #[← f.e]
+  | NormalExpr.cons η (NormalExpr.id _) => η.e
+  | NormalExpr.cons η θ => do mkAppM ``CategoryStruct.comp #[← η.e, ← θ.e]
+
+def NormalExpr.toList : NormalExpr → List WhiskerLeftExpr
+  | NormalExpr.id _ => []
+  | NormalExpr.cons η θ => η :: NormalExpr.toList θ
+
+def WhiskerRightExpr.core : WhiskerRightExpr → Core
+  | WhiskerRightExpr.of η => η
+  | WhiskerRightExpr.whisker η _ => η.core
+
+def WhiskerLeftExpr.core : WhiskerLeftExpr → Core
+  | WhiskerLeftExpr.of η => η.core
+  | WhiskerLeftExpr.whisker _ η => η.core
+
+def WhiskerRightExpr.structural? (η : WhiskerRightExpr) : Option WhiskerRightExpr :=
+  match η.core with
+  | .of _ => none
+  | .ofStructural _ => some η
+
+def WhiskerLeftExpr.structural? (η : WhiskerLeftExpr) : Option WhiskerLeftExpr :=
+  match η.core with
+  | .of _ => none
+  | .ofStructural _ => some η
+
+def NormalExpr.structural? : NormalExpr → Option NormalExpr
+  | NormalExpr.id f => some (.id f)
+  | NormalExpr.cons η θ =>
+    match η.structural?, θ.structural? with
+    | some _, some _ => some (.cons η θ)
+    | _, _ => none
+
+/-- Interpret an `Expr` term as a `Core` term. -/
+def toCore (e : Expr) : Core :=
+  match structuralAtom? e with
+  | some η => Core.ofStructural η
+  | none => Core.of ⟨e⟩
+
+def NormalExpr.of (η : Expr) : MetaM NormalExpr := do
+  return .cons (.of (.of (toCore η))) (.id <| ← StringDiagram.tar η)
+
+/-- Evaluate the expression of a 2-morphism into a normalized form. -/
+partial def eval (e : Expr) : MetaM NormalExpr := do
+  match e.getAppFnArgs with
+  | (``CategoryStruct.id, #[_, _, f]) => return NormalExpr.id (toMor₁ f)
+  | (``CategoryStruct.comp, #[_, _, _, _, _, η, θ]) => return evalComp (← eval η) (← eval θ)
+  | (``MonoidalCategoryStruct.whiskerLeft, #[_, _, _, f, _, _, η]) => evalWhiskerLeftExpr (toMor₁ f) (← eval η)
+  | (``MonoidalCategoryStruct.whiskerRight, #[_, _, _, _, _, η, h]) => evalWhiskerRightExpr (← eval η) (toMor₁ h)
+  | _ => NormalExpr.of e
```

Comment:
I had a go at Qq-ifying this, but the best I could come up with was
```
/-- Evaluate the expression of a 2-morphism into a normalized form. -/
partial def eval {u v : Level} {α : Q(Type u)} {C : Q(Category.{v} $α)} {M : Q(MonoidalCategory $α)}
    {x y : Q($α)} (e : Q($x ⟶ $y)) : MetaM NormalExpr := do
  if let .defEq (h : $x =Q $y) ← isDefEqQ x y then
    if let ~q(𝟙 _) := e then
      return NormalExpr.id (toMor₁ x)
  if let ~q($f ≫ $g) := e then
    return evalComp (← eval f) (← eval g)
  if let ~q($xl ⊗ $xr) := x then
    if let ~q($yl ⊗ $yr) := y then
      if let .defEq (h : $xl =Q $yl) ← isDefEqQ xl yl then
        if let ~q(_ ◁ $η) := e then
          evalWhiskerLeftExpr (toMor₁ xl) (← eval η)
      if let .defEq (h : $xr =Q $yr) ← isDefEqQ xr yr then
        if let ~q($η ▷ _) := e then
          evalWhiskerRightExpr (← eval η) (toMor₁ xr)
  return NormalExpr.of e
  ```
  which doesn't actually work thanks to https://github.com/leanprover-community/quote4/issues/21. So I think Qq-ification is probably not on the cards after all.

Reply:
Thank you very much for trying it! I knew very little about how to use Qq.